ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java

```diff
@@ -471,6 +472,87 @@ public void setAppProperties(@Nullable Bundle appProperties) {
     }
   }
 
+  private static class RootViewInsets {
+    float top;
+    float right;
+    float bottom;
+    float left;
+
+    RootViewInsets(float top, float right, float bottom, float left) {
+      this.top = top;
+      this.right = right;
+      this.bottom = bottom;
+      this.left = left;
+    }
+  }
+
+  private RootViewInsets getSafeAreaInsets() {
+    // Window insets are parts of the window that are covered by system views (status bar,
+    // navigation bar, notches). There are no apis the get these values for android < M so we
+    // do a best effort polyfill.
+    // TODO: Use the DisplayCutout api when we target sdk 28+
+    RootViewInsets windowInsets;
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      WindowInsets insets = getRootWindowInsets();
+      windowInsets = new RootViewInsets(
+        insets.getSystemWindowInsetTop(),
+        insets.getSystemWindowInsetRight(),
+        insets.getSystemWindowInsetBottom(),
+        insets.getSystemWindowInsetLeft());
+    } else {
+      int rotation =
+        ((WindowManager) getContext().getSystemService(Context.WINDOW_SERVICE))
+          .getDefaultDisplay().getRotation();
+      int statusBarHeight = 0;
+      int resourceId = getResources().getIdentifier("status_bar_height", "dimen", "android");
+      if (resourceId > 0) {
+        statusBarHeight = getResources().getDimensionPixelSize(resourceId);
+      }
+      int navbarHeight = 0;
+      resourceId = getResources().getIdentifier("navigation_bar_height", "dimen", "android");
+      if (resourceId > 0) {
+        navbarHeight = getResources().getDimensionPixelSize(resourceId);
+      }
+
+      windowInsets = new RootViewInsets(
+        statusBarHeight,
+        rotation == Surface.ROTATION_90 ? navbarHeight : 0,
+        rotation == Surface.ROTATION_0 || rotation == Surface.ROTATION_180 ? navbarHeight : 0,
```

Comment:
Apparently some tablet devices can render navigation bar on the bottom while reporting landscape rotation 😢  Maybe we can use this method as a polyfil -> https://developer.android.com/reference/android/view/View.html#getWindowVisibleDisplayFrame(android.graphics.Rect)

Reply:
Ew, I'll look into this method to see if I can use this for the polyfill instead of using status bar / navigation bar dimensions.